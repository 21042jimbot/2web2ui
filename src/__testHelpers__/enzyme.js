export const findButton = (wrapper, buttonLabel) =>
  wrapper.findWhere(
    (node) => node.type() === 'button' && new RegExp(buttonLabel).test(node.text())
  );
