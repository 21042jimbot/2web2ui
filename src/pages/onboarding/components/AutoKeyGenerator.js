import React, { Component } from 'react';
import { withRouter } from 'react-router';
import PropTypes from 'prop-types';
import { showAlert } from 'src/actions/globalAlert';
import { createApiKey, hideNewApiKey } from 'src/actions/api-keys';
import { connect } from 'react-redux';
import { Loading } from 'src/components';

const keyDefaults = {
  label: 'Send Email Key (auto-generated)',
  grants: ['smtp/inject', 'transmissions/modify', 'templates/view']
};

// Handles API key generation
// Passes down newly created apiKey and current user's email address
export class AutoKeyGenerator extends Component {
  componentDidMount() {
    this.createKey();
  }

  componentWillUnmount() {
    this.props.hideNewApiKey();
  }

  createKey = () => {
    const { createApiKey, showAlert, history } = this.props;

    return createApiKey(keyDefaults).catch((err) => {
      showAlert({ type: 'error', message: 'Sorry, something went wrong', details: err.message });
      history.push('/dashboard');
    });
  }

  render() {
    const { apiKey, render, email } = this.props;

    if (!apiKey) {
      return <Loading />;
    }

    return render({ apiKey, email });
  }
}

AutoKeyGenerator.defaultProps = {
  email: 'email address here'
};

AutoKeyGenerator.propTypes = {
  render: PropTypes.func.isRequired,
  apiKey: PropTypes.string,
  email: PropTypes.string
};

const mapStateToProps = ({ apiKeys, currentUser }) => ({
  apiKey: apiKeys.newKey,
  email: currentUser.email
});
export default withRouter(connect(mapStateToProps, { createApiKey, hideNewApiKey, showAlert })(AutoKeyGenerator));
