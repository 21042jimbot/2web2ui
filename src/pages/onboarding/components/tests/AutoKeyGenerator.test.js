import { shallow } from 'enzyme';
import React from 'react';
import { AutoKeyGenerator } from '../AutoKeyGenerator';

const keyDefaults = {
  label: 'Send Email Key (auto-generated)',
  grants: ['smtp/inject', 'transmissions/modify', 'templates/view']
};

describe('AutoKeyGenerator', () => {
  let wrapper;

  beforeEach(() => {
    const props = {
      createApiKey: jest.fn(() => Promise.resolve()),
      hideNewApiKey: jest.fn(),
      history: {
        push: jest.fn()
      },
      render: jest.fn(() => 'Rendered Component'),
      apiKey: '123456789'
    };

    wrapper = shallow(<AutoKeyGenerator {...props} />);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should create a new key on mount', () => {
    expect(wrapper.instance().props.createApiKey).toHaveBeenCalledWith(keyDefaults);
  });

  it('should handle failure on new key creation', async() => {
    const createFailure = jest.fn(() => Promise.reject({ message: 'error' }));
    wrapper.setProps({ createApiKey: createFailure });
    await wrapper.instance().createKey();

    const props = wrapper.instance().props;
    expect(createFailure).toHaveBeenCalledWith(keyDefaults);
    expect(props.history.push).toHaveBeenCalledWith('/dashboard');
  });

  it('should render loading without an api key', () => {
    wrapper.setProps({ apiKey: null });
    expect(wrapper).toMatchSnapshot();
  });

  it('should call render with api key and default email', () => {
    const props = wrapper.instance().props;
    expect(props.render).toHaveBeenCalledWith({ apiKey: props.apiKey, email: 'email address here' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should call render with api key and user email', () => {
    wrapper.setProps({ email: 'user@email.com' });
    const props = wrapper.instance().props;
    expect(props.render).toHaveBeenCalledWith({ apiKey: props.apiKey, email: 'user@email.com' });
  });

  it('should clear new key on unmount', () => {
    const props = wrapper.instance().props;
    wrapper.unmount();
    expect(props.hideNewApiKey).toHaveBeenCalled();
  });
});
