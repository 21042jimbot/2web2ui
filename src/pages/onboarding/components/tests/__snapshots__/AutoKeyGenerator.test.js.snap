// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoKeyGenerator should call render with api key and default email 1`] = `"Rendered Component"`;

exports[`AutoKeyGenerator should render loading without an api key 1`] = `<Loading />`;
